src/_P021_Level.ino

```diff
@@ -18,39 +22,179 @@
+# define P021_GUID_MIN_TIME         "f9"
+
+// Maximum time output may be idle [sec]/[hour]
+# define P021_INTERVAL_TIME         PCONFIG_ULONG(2)
+# define P021_GUID_INTERVAL_TIME    "f10"
+
+// Duration for forced circulation [sec]/[min]
+# define P021_FORCE_TIME            PCONFIG_ULONG(3)
+# define P021_GUID_FORCE_TIME       "f11"
+
+// Positions in userVar array for the output values of this plugin
+// For now we only advertise output & state. Autosave bookkeeping is hidden and used as static storage
+#define P021_VALUE_OUTPUT           0 // Switch output, logical state [inactive/active]
+#define P021_VALUE_STATE            1 // Control state, see P021_control_state
+#define P021_VALUE_AUTOSAVE_FLAG    2 // Autosave bookkeeping
+#define P021_VALUE_AUTOSAVE_TIME    3 // Autosave bookkeeping
+
+// Simple conversion from parameter settings in hours/minutes to seconds (administration unit)
+// And from seconds to the millis domain used for the actual control
+// Note that these simple conversion may lose precision due to rough rounding
+
+int millis2seconds(int32_t x) {
+  return (uint)(x / 1000);
+}
+
+int32_t seconds2millis(int x) {
+  return (uint32_t)(x * 1000);
+}
 
-# define P021_CHECK_TASK          PCONFIG(0)
-# define P021_CHECK_VALUE         PCONFIG(1)
-# define P021_DONT_ALWAYS_SAVE    PCONFIG(2)
-# define P021_INVERT_OUTPUT       PCONFIG(3)
-# define P021_TRIGGER_LEVEL       PCONFIG_FLOAT(0)
-# define P021_TRIGGER_HYSTERESIS  PCONFIG_FLOAT(1)
-# define P021_TRIGGER_LAST_STORED PCONFIG_FLOAT(2)
-# define P021_AUTOSAVE_TIMER      PCONFIG_ULONG(0)
+# ifndef P021_MIN_BUILD_SIZE
+int minutes2seconds(int x) {
+  return x * 60;
+}
+
+int hours2seconds(int x) {
+  return x * 60 * 60;
+}
+
+int seconds2minutes(int x) {
+  return x / 60;
+}
 
+int seconds2hours(int x) {
+  return x / (60 * 60);
+}
+
+# endif // ifndef P021_MIN_BUILD_SIZE
+
+// Operation modes for the control algorithm
+enum P021_opmode
+{
+  P021_OPMODE_CLASSIC, // Original, stateless control
+  P021_OPMODE_OFF,     // Output is fully shut down, no forced curculation
+  P021_OPMODE_STANDBY, // Output is only switched on for forced maintenance runs
+  P021_OPMODE_ON,      // Output is always switched on
+  P021_OPMODE_TEMP,    // Control algorithm based on temperature only
+  P021_OPMODE_REMOTE   // Both temperature and remote command can switch on Output
+};
+
+// Control state for the control algorithm
+enum P021_control_state
+{
+  P021_STATE_IDLE,   // Output is inactive
+  P021_STATE_ACTIVE, // Output is active due to level control (based on input value)
+  P021_STATE_EXTEND, // Output is active due to minimum duration
+  P021_STATE_FORCE   // Output is forced active due to maximum inactive duration is exceeded
+};
+
+// Static storage for global state info. Track per ESPeasy plugin instance
+static bool P021_remote[TASKS_MAX];        // Static storage for remote control.
+static uint32_t P021_timestamp[TASKS_MAX]; // Static storage for timestamp last change
```

Comment:
Why not use a PluginStruct to store these values?
That's why I created those, so you don't need to allocate these kind of arrays which are hardly ever used but still take up precious memory.

A few very simple and small examples: 
- P024_data_struct.h/.cpp
- P060_data_struct.h/.cpp
- P069_data_struct.h/.cpp
- P107_data_struct.h/.cpp

Reply:
I know these structures. I copied the pattern from the original plugin. Seems we are lacking flash memory more than ram memory and adding a complete class probably requires more code in flash. But I will look at it. Both variables don't show on the webform. They are initialized when the plugin is initialized and used to track data between calls.

Reply:
Sorry, timer initialization got lost during the optimizations.